File: agenkit-scala/src/main/scala/io/agenkit/patterns/ReasoningWithToolsAgent.scala
package io.agenkit.patterns

import io.agenkit.adapters.LlmClient
import io.agenkit.core.*
import scala.concurrent.{ExecutionContext, Future}
import java.util.concurrent.atomic.AtomicLong

/** Step-by-step reasoning agent that can invoke tools between reasoning steps. */
class ReasoningWithToolsAgent(
  val name: String,
  llm: LlmClient,
  tools: Map[String, Tool] = Map.empty,
  maxSteps: Int = 10
) extends Agent:
  private val _processedCount = new AtomicLong(0)

  def capabilities: List[String] = List("reasoning", "tool-use", "step-by-step")

  def process(message: Message)(using ExecutionContext): Future[Message] =
    _processedCount.incrementAndGet()
    val toolDescs = tools.values.map(t => s"${t.name}: ${t.description}").mkString("\n")
    val systemMsg = Message.system(
      s"""Reason step by step.  Use:
         |THINK: your reasoning
         |USE_TOOL: tool_name with params key=value
         |ANSWER: final answer
         |
         |Tools: $toolDescs""".stripMargin
    )
    reasoningLoop(List(systemMsg, message), 0)

  private def reasoningLoop(
    context: List[Message],
    step: Int
  )(using ExecutionContext): Future[Message] =
    if step >= maxSteps then
      Future.successful(Message.of("assistant", "Max reasoning steps reached"))
    else
      llm.complete(context).flatMap { response =>
        val content = response.contentString
        if content.contains("ANSWER:") then
          val answer = content.split("ANSWER:").last.trim
          Future.successful(Message.of("assistant", answer))
        else if content.contains("USE_TOOL:") then
          val toolLine = content.split("\n").find(_.contains("USE_TOOL:")).getOrElse("")
          val parts    = toolLine.stripPrefix("USE_TOOL:").trim.split("\\s+with\\s+params\\s+", 2)
          val toolName = if parts.nonEmpty then parts(0).trim else ""
          val params   = if parts.length > 1 then parseParams(parts(1)) else Map.empty[String, Any]
          tools.get(toolName) match
            case Some(tool) =>
              tool.execute(params).flatMap { result =>
                val obs = result match
                  case ToolOk(data)    => s"TOOL_RESULT: $data"
                  case ToolFail(error) => s"TOOL_RESULT: Error: $error"
                reasoningLoop(context ++ List(response, Message.user(obs)), step + 1)
              }
            case None =>
              val obs = s"TOOL_RESULT: Tool '$toolName' not found"
              reasoningLoop(context ++ List(response, Message.user(obs)), step + 1)
        else
          reasoningLoop(context :+ response, step + 1)
      }

  private def parseParams(str: String): Map[String, Any] =
    str.split(",").flatMap { p =>
      p.split("=", 2) match
        case Array(k, v) => Some(k.trim -> (v.trim: Any))
        case _           => None
    }.toMap

  def introspect(): IntrospectionResult =
    IntrospectionResult(
      name = name,
      metadata = Map("tools" -> tools.keys.toList, "maxSteps" -> maxSteps),
      capabilities = capabilities,
      processedMessages = _processedCount.get()
    )

File: agenkit-java/src/main/java/io/agenkit/patterns/ReasoningWithToolsAgent.java
package io.agenkit.patterns;

import io.agenkit.adapters.LlmClient;
import io.agenkit.core.Agent;
import io.agenkit.core.IntrospectionResult;
import io.agenkit.core.Message;
import io.agenkit.core.Tool;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;

/**
 * An agent that interleaves step-by-step reasoning with tool invocations.
 */
public final class ReasoningWithToolsAgent implements Agent {

    private final String name;
    private final LlmClient llmClient;
    private final List<Tool> tools;
    private final int maxSteps;

    public ReasoningWithToolsAgent(String name, LlmClient llmClient, List<Tool> tools, int maxSteps) {
        this.name = name;
        this.llmClient = llmClient;
        this.tools = List.copyOf(tools);
        this.maxSteps = maxSteps;
    }

    public ReasoningWithToolsAgent(String name, LlmClient llmClient, List<Tool> tools) {
        this(name, llmClient, tools, 8);
    }

    @Override
    public String getName() { return name; }

    @Override
    public List<String> getCapabilities() {
        return List.of("reasoning", "tool_use", "chain_of_thought");
    }

    @Override
    public CompletableFuture<Message> process(Message message) {
        return reason(message, new ArrayList<>(), 0);
    }

    private CompletableFuture<Message> reason(Message original, List<String> chain, int step) {
        if (step >= maxSteps) {
            return CompletableFuture.completedFuture(
                    Message.of("assistant", String.join("\n", chain))
                            .withMetadata("steps", step)
                            .withMetadata("stopped", "max_steps"));
        }

        StringBuilder toolsDesc = new StringBuilder();
        for (Tool tool : tools) {
            toolsDesc.append("- ").append(tool.getName())
                    .append(": ").append(tool.getDescription()).append("\n");
        }

        String context = "Question: " + original.contentString()
                + (chain.isEmpty() ? "" : "\n\nReasoning so far:\n" + String.join("\n", chain));

        return llmClient.complete(List.of(
                Message.of("system", "Reason step by step. Available tools:\n" + toolsDesc
                        + "\nFormat: THINK: reasoning\nUSE_TOOL: tool_name INPUT: input\nOR: ANSWER: final answer"),
                Message.of("user", context)))
                .thenCompose(response -> {
                    String text = response.contentString();

                    if (text.contains("ANSWER:")) {
                        String answer = text.substring(text.indexOf("ANSWER:") + 7).trim();
                        return CompletableFuture.completedFuture(
                                Message.of("assistant", answer)
                                        .withMetadata("steps", step + 1)
                                        .withMetadata("reasoning_chain", chain.size()));
                    }

                    if (text.contains("USE_TOOL:")) {
                        String toolPart = text.substring(text.indexOf("USE_TOOL:") + 9);
                        String toolName = toolPart.split("INPUT:")[0].trim();
                        String input = toolPart.contains("INPUT:")
                                ? toolPart.substring(toolPart.indexOf("INPUT:") + 6).trim() : "";

                        Tool tool = tools.stream()
                                .filter(t -> t.getName().equalsIgnoreCase(toolName))
                                .findFirst().orElse(null);

                        if (tool != null) {
                            Map<String, Object> params = new HashMap<>();
                            params.put("input", input);
                            return tool.execute(params).thenCompose(result -> {
                                List<String> newChain = new ArrayList<>(chain);
                                newChain.add(text);
                                newChain.add("Tool result: " + result.getData());
                                return reason(original, newChain, step + 1);
                            });
                        }
                    }

                    List<String> newChain = new ArrayList<>(chain);
                    newChain.add(text);
                    return reason(original, newChain, step + 1);
                });
    }

    @Override
    public IntrospectionResult introspect() {
        List<String> toolNames = tools.stream().map(Tool::getName).toList();
        Map<String, Object> state = new HashMap<>();
        state.put("maxSteps", maxSteps);
        return new IntrospectionResult(name, getCapabilities(), null, state, toolNames);
    }
}
